pontoon/base/models.py

```diff
@@ -284,53 +284,82 @@ def latest_activity(self, locale=None):
 
         return translations.latest_activity()
 
+    def add_detail(self, array, path, entity, translated, approved):
+        return array.append({
+            'resource__path': path,
+            'resource__entity_count': entity,
+            'translated_count': translated,
+            'approved_count': approved
+        })
+
     @property
     def locales_parts_stats(self):
         """Get project locales with their pages/paths and stats."""
-        pages = self.subpage_set.all()
-        resources = self.resource_set.filter(entity__obsolete=False).distinct()
         details = {}
 
-        for locale in self.locales.all():
-            stats = Stats.objects.filter(resource__in=resources, locale=locale)
+        # Project Locales with prefetched Stats and Resources
+        qs = (
+            Stats.objects
+            .select_related('resource')
+            .filter(resource__project=self, resource__entity__obsolete=False)
+            .distinct()
+        )
+        locales = self.locales.prefetch_related(
+            Prefetch(
+                'stats_set',
+                queryset=qs,
+                to_attr='fetched_stats'
+            )
+        )
+
+        # Project Pages with prefetched Resources
+        pages = self.subpage_set.prefetch_related(
+            'resources'
+        )
+
+        has_pages = pages.exists()
+
+        for locale in locales:
             locale_details = []
 
-            # Get subpage name and stats if subpages defined
-            if pages.exists():
+            # If project has defined subpages, return their names with
+            # corresponding project stats. If subpages have defined resources,
+            # only include stats for page resources.
+            if has_pages:
 
                 # List only subpages, whose resources are available for locale
-                pages = pages.filter(resources__stats=stats) or pages
+                pages = pages.filter(resources__stats__in=locale.fetched_stats) or pages
 
                 for page in pages:
+                    entity_count = translated_count = approved_count = 0
```

Comment:
If you want to avoid having three lines here, don't rely on assignment returning a value. Use destructuring instead:

```python
entity_count, translated_count, approved_count = 0, 0, 0
```